File: src/8-cpp-class-wrap/8-1/1/mybuffer.cpp
#include "mybuffer.h"

MyBuffer::MyBuffer(int size) {
  this->s_ = new std::string(size, char('\0'));
}

MyBuffer::~MyBuffer() {
  delete this->s_;
}

int MyBuffer::Size() const{
  return this->s_->size();
}

char* MyBuffer::Data() {
  return (char*)this->s_->c_str();
}

File: src/8-cpp-class-wrap/8-1/3/mybuffer/mybuffer.go
package mybuffer

import "unsafe"

// MyBuffer ...
type MyBuffer struct {
	cptr *cgoMyBufferT
}

// NewMyBuffer ...
func NewMyBuffer(size int) *MyBuffer {
	return &MyBuffer{
		cptr: cgoNewMyBuffer(size),
	}
}

// Delete ...
func (p *MyBuffer) Delete() {
	cgoDeleteMyBuffer(p.cptr)
}

// Data ...
func (p *MyBuffer) Data() []byte {
	data := cgoMyBufferData(p.cptr)
	size := cgoMyBufferSize(p.cptr)
	return ((*[1 << 31]byte)(unsafe.Pointer(data)))[0:int(size):int(size)]
}
